Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs

```diff
@@ -69,187 +70,203 @@ public override void OnEpisodeBegin()
-                    // otherwise it's hitting an obstacle or a different enemy and is false
-                    if (hit.transform.gameObject == enemy)
-					{
-						los = true;
-						// Debug.DrawRay(startpt, (enemy.transform.position - transform.position), Color.green);
-					}
-					// else
-					// {
-					// 	Debug.DrawRay(startpt, (enemy.transform.position - transform.position), Color.red);
-					// }
-                }
-				float normalizedLos = los ? 1.0f : 0.0f;
-
-				if (enemyNormalizedHealth <= 0.0f)
-				{
-					// if they're dead, zero out observations before saving/sending them below
-					normalizedDistance = 0f;
-					normalizedRelativeBearing = 0f;
-					normalizedHeading = 0f;
-					enemyNormalizedHealth = 0f;
-					normalizedLos = 0f;
-				}
-				else
-				{
-					// if not dead, add their angle to the list of enemy angles to use later -- we're only adding live enemies angles here
-					enemyAngles.Add(Mathf.Abs(normalizedRelativeBearing));
-				}
+			for (int i = 0; i < enemies.Count; i++)
+			{
+				var enemy = enemies[i];
+				if (!enemy) continue;
+
+				float[] enemyObservations = GetEnemyObservations(sensor, enemy, i);
+				enemyBufferSensor.AppendObservation(enemyObservations);
 
-				float[] enemyObs = { normalizedDistance, normalizedRelativeBearing, normalizedHeading, enemyNormalizedHealth, normalizedLos };
-				enemyBufferSensor.AppendObservation(enemyObs);
 				// since there's no easy way to query buffer sensor for obs, we need to manually track below, for debug display / testing
 				debugEnemyBufferSensorObs.Add(string.Format("Enemy[{0}]\nDistance: {1}\nRelBearing: {2}\nHeading: {3}\nHealth: {4}\nLOS: {5}",
-					i, normalizedDistance, normalizedRelativeBearing, normalizedHeading, enemyNormalizedHealth, normalizedLos));
+					i, enemyObservations[0], enemyObservations[1], enemyObservations[2], enemyObservations[3], enemyObservations[4]));
 
 				// pants on fire reward
 				AddReward(-1.0f / MaxStep);
-            }
+			}
+		}
+
+		private float[] GetEnemyObservations(VectorSensor sensor, GameObject enemy, int enemyIndex)
+		{
+			float distance = 0f;
+			float bearing = 0f;
+			float heading = 0f;
+			float lineOfSight = 0f;
+			float health = ObserveEnemyHealth(sensor, enemy, enemyIndex);
+
+			if (health > 0.0f)
+			{
+				distance = DetermineDistanceFrom(enemy);
+				bearing = DetermineBearingTo(enemy);
+				heading = DetermineHeadingOf(enemy);
+				lineOfSight = DoesAgentHaveLineOfSightOf(enemy);
+				enemyAngles.Add(Mathf.Abs(bearing));
+			}
+			else
+			{
+				health = 0f;
+			}
+			float[] observations = { distance, bearing, heading, health, lineOfSight };
+			return observations;
+		}
+
+		private float ObserveEnemyHealth(VectorSensor sensor, GameObject enemy, int enemyIndex)
+		{
+			var enemyHealthComponent = enemy.GetComponent<Unity.FPS.Game.Health>();
+			float enemyNormalizedHealth = enemyHealthComponent.GetRatio();
```

Comment:
we should put in a conditional to make sure we dont have a null Health component.  if null, then it can skip and use 0

Reply:
Fixed! you null it 🤩